### DISCUSSION TITLE
Help for high-dimensional dataset

### DISCUSSION THREAD

**Discussion by marekdetven** [MilesCranmer/PySR#948] [Q&A] [answered]
Hello! I am trying to use PySR to find a complex cosmological function, which depends on 16 input variables. I have been running PySR on my dataset for 2-3 days, and I have gotten down to a MSE of 37.8. I would like to get the MSE far lower, but I recently read that genetic algorithms like PySR are not adapted to fitting functions with this many input parameters. My dataset has about 7,500 points. Here is the model that I am using:
```
    parallelism     = "multithreading",
    model_selection = "best",
    niterations     = 5000,
    warm_start      = True,
    populations     = 10,
    population_size = 300,
    maxsize         = 150,
    logger_spec     = logger_spec,
    verbosity       = 0,
    binary_operators = ["+","*","^","-","/"],
    unary_operators  = [
        "safe_sin(x) = isfinite(x) ? sin(x) : zero(x)",
        "tanh", "abs", "atan",
        "safe_log(x) = (x > 0) ? log(x) : zero(x)",
        "exp", "cosh", "sinh"
    ],
    output_directory="outputs/models",
    run_id          = runid,
    constraints        = {"^": (-1, 2)},
    nested_constraints = {
        op: {"safe_sin":1,"tanh":1,"sinh":1,"atan":1,"cosh":1}
        for op in ["safe_sin","tanh","sinh","atan","cosh"]
    },
    parsimony       = 0.0005,
    weight_optimize = 0.25,
    batching    = True,
    batch_size  = 1000,
    fast_cycle  = True,
    turbo       = True,
    extra_sympy_mappings = {
        "safe_sin": lambda x: sin(x),
        "safe_log": lambda x: log(x),
    },
```

Do any of the settings that I am using jump out as something that could harm the quality of my runs? Are there things that I should maybe think about changing? Or do I just need to let it run for really long?

I also found the article on symbolic deep learning to be interesting in the context of a function like mine which has so many input dimensions. Am I better off trying to first work with the deep learning model rather than to directly run PySR on this 16-input function? What is roughly the max number of input variables that PySR can take and still perform well with?

Thanks a lot for the help.

**Comment by MilesCranmer** [comment]
Just to check, do you have 7,500 features/columns, or do you mean 7,500 rows?

Be sure to check the advice here: https://ai.damtp.cam.ac.uk/pysr/tuning/

**Reply by marekdetven** [reply]
I have 16 features, and 7,500 rows. Sorry for the confusion!

**Reply by MilesCranmer** [reply]
16 features should be fine. That many rows is fine too. But you could try a smaller batch size. Maybe like `64`?

**Reply by marekdetven** [reply]
Perfect, I will try that. I am guessing that it will significantly speed up the exploration?

**Comment by MilesCranmer** [comment]
By the way, you can just use regular `log` and `sin`. They have safe versions internally anyways

**Comment by MilesCranmer** [comment, marked as answer]
Other advice:

- Turn off most of those unary operators. Generally you want to avoid having degenerate operators. For example, exp and `^` are degenerate. As are sinh, cosh, and tanh. So you could remove those ones.
  - Also I feel like `abs` probably is not physical, so you could turn that off too?
- Depending on your dataset, it might be easier if you pass the logarithm of certain features. For example if you expect a power law, it's easier to fit it in log space from the start.
- Consider other loss functions based on the physics you are fitting.
- Consider a template expression, if you already have a good guess for the functional form.
- Much smaller batch size
- Turn down weight_optimize; that one is quite large and would result in many steps doing BFGS optimization, which is very expensive.
- Generally for most hyperparameters, you are often better sticking with the default unless you have tuned it and found it to work better. For example, the `populations` you have is quite small. Is that based on tuning it?
- The maxsize is a little bit large. Maybe you need it because of the higher number of features (?) but you could try a lower value just to see. Also maybe try the option related to warming up maxsize.
- fast_cycle is deprecated. I guess maybe there is a docs page somewhere that suggests it? (Could you let me know where you read it, so I can go update the docs)

**Reply by marekdetven** [reply]
Thanks for all of the advice - the choice of populations was not based on tuning, so I will just stick to the default values. 
Indeed, to incorporate features which impact the function less than others I have found this maxsize to be good but I might test some lower values. 
I got the fast_cycle from the API reference https://ai.damtp.cam.ac.uk/pysr/api/

**Comment by MilesCranmer** [comment]
I'd also recommend fitting some other ML model (XGBoost or random forest) to quickly check what a "good" loss is. This would tell you: Is a lower MSE actually achievable? If the model does get a lower MSE, does that hurt generalisation to some part of the space of physical parameters? This would tell you what bar to aim for here.

**Comment by gm89uk** [comment]
I excluded some variables when running SR. I then ran XGBoost with crossvalidation to predict the residuals from the SR model. The SHAP values of the XGBoost model will then give you an an idea if any of the remaining variables carry any useful predictability.

**Reply by marekdetven** [reply]
Thank you for the advice ! I will try to do that as well.
